### ISSUE TITLE
Weird formatting for method names

### ISSUE THREAD

**Issue by sander1095** [belav/csharpier#997]
Hi! 

I've written the following code which I find to be very readable:

```csharp
[HttpGet("dosinglanes")]
public async Task<ActionResult<IReadOnlyCollection<GetProductListResponseModel>>> GetProductsOnDosingLanes()
{
    await Task.CompletedTask;
    return Ok();
}
```

However, `csharpier` turns this into the following:

```csharp
[HttpGet("dosinglanes")]
public async Task<
    ActionResult<IReadOnlyCollection<GetProductListResponseModel[]>>
> GetProductsOnDosingLanes()
{
    await Task.CompletedTask;
    return Ok();
}
```

Which I find to be very **unreadable**. The `> GetProductsOnDosingLanes()` is a very.. unique way of formatting this code.

I would like Csharpier to keep the code as is. What are your thoughts on this?
I could simply add a `// csharpier-ignore`, but I don't think this code should be formatted :)

**Comment by sander1095** [comment]
I've implemented a temporary workaround by using `.csharpierrc` like this:

```
# Some long API method definitions are easier to read when on one line.
printWidth: 120
```

But this is not ideal for me, because the "default" print width is fine by me, except for this case.
I know formatting is a never-ending discussion, so feel free to close this issue if you believe this is issue is a matter of preference and the `.csharpierrc` is a good enough fix. Otherwise, let's have a discussion :D

**Comment by belav** [comment]
> Perhaps it would be possible to apply .csharpierrc rules for specific files/directories, including wildcards? so like /**/*Controller.cs would only have a longer printWidth.

I believe you can make that work with `.editorconfig`. As of 0.26.0 csharpier supports reading editorconfig files.

I agree that the method isn't formatted very well. Maybe something like this?

```c#
[HttpGet("dosinglanes")]
public async 
    Task<ActionResult<IReadOnlyCollection<GetProductListResponseModel>>> 
    GetProductsOnDosingLanes()
{
```

Or if there is only a single generic parameter never break, which I believe would result in this no matter how long the method name gets
```c#
[HttpGet("dosinglanes")]
public async Task<ActionResult<IReadOnlyCollection<GetProductListResponseModel>>> GetProductsOnDosingLanes()
{
```

I'll have to play around a bit with other sample code to see how both of those end up, or if there are any other possiblities.

**Comment by jods4** [comment]
Current formatting in OP breaks the Rectangle rule, and it doesn't feel good in the middle of a generic type.

> Maybe something like this?
```cs
public async 
    Task<ActionResult<IReadOnlyCollection<GetProductListResponseModel>>> 
    GetProductsOnDosingLanes()
{
```

Seems ok, altough maybe I'd keep the return type on first line:
```cs
public async Task<ActionResult<IReadOnlyCollection<GetProductListResponseModel>>> 
    GetProductsOnDosingLanes()
{
```

This made me think of a new concept that could solve many of csharpier not-so-pretty edge cases. It'll need more thinking and design but it has potential:
_CSharpier should avoid breaks that create widows or orphans._

I borrow these terms from typography, what I mean for code is: lines that are very short, so the break doesn't actually save much.

In this example, leaving `public async` alone on a line is only 12 characters. Maybe that's not worthy of breaking and the return type should be left on the same line.
A similar argument could be made for the method name (so keep everything on one line as you suggest), although 26 characters might be enough to justify the break.
As a counter-example, if the method was named `Apply()`, even if the line was over the limit it would be silly to break it on its own line.

This concept could help here, and many other instances.
This example from #672 has many lines with 3 or 5 characters: 
```cs
class ClassName
{
    void MethodName()
    {
        var searchRequest = new SearchDescriptor<ElasticsearchProduct>().WithAggregations(
            o =>
                o.WithFilter(
                    key,
                    f =>
                        f.WithFilter(fd => filter)
                            .WithAggregations(
                                ad =>
                                    ad.WithTerm(
                                        key,
                                        cd =>
                                            cd.WithField(facetField).WithSize(MaximumCategoryFacets)
                                    )
                            )
                )
        );
    }
}
```

Let's use the same breaking logic, but avoid breaks that leaves less than 10 chars of syntax on a single line (also adjusted the chaining to match more recent CSharpier releases):
```cs
class ClassName
{
    void MethodName()
    {
        var searchRequest = new SearchDescriptor<ElasticsearchProduct>()
            .WithAggregations(o =>
                o.WithFilter(
                    key,
                    f => f
                        .WithFilter(fd => filter)
                        .WithAggregations(ad => 
                            ad.WithTerm(
                                key, 
                                cd => cd.WithField(facetField).WithSize(MaximumCategoryFacets)
                            )
                        )
                )
            )
        );
    }
}
```
I think it's an improvement.
Note that I choose not to apply a minimum length rule to one argument in a list of arguments or to one method call in a chain of calls. I think that would hurt readability.

**Comment by belav** [comment]
> Let's use the same breaking logic, but avoid breaks that leaves less than 10 chars of syntax on a single line:

I'm really curious if there may be an easy way to implement this. Although it probably needs to be applied selectively. Some of the parameters in your example are less than 10 chars. But those are part of a group, where `o =>` is not. Definitely something worth exploring more!

**Comment by vlflorian** [comment]
Has anyone found a workaround or fix for this?  

Also encountering really weird method formatting here but can't find a config option to disable it.

```c#
public async Task<
        PagedResult<LegalProcedureReason, Embedded<LegalProcedureReason>>
    > GetLegalProcedureReasonsAsync(bool bypassCache, CancellationToken token)
    {
```

I'd rather see something like this:

```c#
    public async Task<PagedResult<LegalProcedureReason, Embedded<LegalProcedureReason>>>
	    GetLegalProcedureReasonsAsync(bool bypassCache, CancellationToken token)
    {

```

**Comment by belav** [comment]
Some related issues #450 #468 

Another example of a method
```c#
static async Task<
	Results<Ok<PaginatedResponse<QuestionnaireDisplayModel>>, ErroredProblem>
> AllQuestionnairesAsync(
	[AsParameters] PaginationRequest<IQuestionnaireService> request,
	CancellationToken cancellationToken
)
```

FWIW this is how prettier deals with similar generics in TS. I haven't looked to see if they have any discussions on better ways to deal with it.
```TS
function AllQuestionnairesAsync(
  request: Task<
    Results<Ok<PaginatedResponse<QuestionnaireDisplayModel>>, ErroredProblem>
  >,
  cancellationToken: CancellationToken,
): Task<
  Results<Ok<PaginatedResponse<QuestionnaireDisplayModel>>, ErroredProblem>
> {
  return true;
}

**Comment by belav** [comment]
@krtek2k this issue exists because we all agree that there is improvement to be made. Calling someone delusional doesn't really add to this discussion.

**Comment by krtek2k** [comment]
so nothing, closing duplicates, years passes by
can you please share the workaround because it is a deal breaker for me, and all other formatting works well and is pretty neat, so I would like to join the show, maybe introduce some kind of advanced settings? The options for the extension are pretty limited to nothing..

**Comment by belav** [comment]
> so nothing, closing duplicates, years passes by

Yes. Duplicates get closed, did you have something else in mind that I should do with them?

Yes. In the open source world sometimes it takes years to address things that aren't a high priority.

> can you please share the workaround because it is a deal breaker for me, and all other formatting works well and is pretty neat, so I would like to join the show, maybe introduce some kind of advanced settings? The options for the extension are pretty limited to nothing..

CSharpier will not add any more options. 

As a workaround you can ignore the line with `// csharpier-ignore` and format it yourself. 

Alternatively you could clone this repo, come up with some better way to format generics that handles all of the edge cases while also keeping the userbase happy. Then submit a PR with those changes.

**Comment by krtek2k** [comment]
Thanks for your time.
